onnxruntime/core/framework/provider_bridge_ort.cc

```diff
@@ -49,487 +76,553 @@ struct Provider_IDeviceAllocator_Impl : Provider_IDeviceAllocator {
+
+  google::protobuf::internal::RepeatedPtrIterator<const onnx::TensorShapeProto_Dimension> v_;
 };
 
-struct Provider_AttributeProto_Impl : ONNX_NAMESPACE::Provider_AttributeProto {
-  Provider_AttributeProto_Impl() = default;
-  Provider_AttributeProto_Impl(const ONNX_NAMESPACE::AttributeProto& copy) : v_{copy} {}
+struct Provider_NodeAttributes_Iterator_Impl : Provider_NodeAttributes_Iterator {
+  Provider_NodeAttributes_Iterator_Impl(NodeAttributes::const_iterator&& v) : v_{std::move(v)} {}
 
-  std::unique_ptr<Provider_AttributeProto> Clone() const override {
-    return onnxruntime::make_unique<Provider_AttributeProto_Impl>(v_);
-  }
+  bool operator!=(const Provider_NodeAttributes_Iterator& p) const override { return v_ != static_cast<const Provider_NodeAttributes_Iterator_Impl*>(&p)->v_; }
 
-  ::onnx::AttributeProto_AttributeType type() const override { return v_.type(); }
+  void operator++() override { v_.operator++(); }
+  const std::string& first() const override { return v_->first; }
+  const Provider_AttributeProto& second() override { return *reinterpret_cast<const Provider_AttributeProto*>(static_cast<const ONNX_NAMESPACE::AttributeProto*>(&v_->second)); }
 
-  int ints_size() const override {
-    return v_.ints_size();
-  }
+  NodeAttributes::const_iterator v_;
+};
 
-  int64_t ints(int i) const override { return v_.ints(i); }
-  int64_t i() const override { return v_.i(); }
-  float f() const override { return v_.f(); }
-  void set_s(const ::std::string& value) override { v_.set_s(value); }
-  const ::std::string& s() const override { return v_.s(); }
-  void set_name(const ::std::string& value) override { v_.set_name(value); }
-  void set_type(::onnx::AttributeProto_AttributeType value) override { v_.set_type(value); }
-  ::onnx::Provider_TensorProto* add_tensors() override {
-    // Kind of a hack, but the pointer is only valid until the next add_tensors call
-    tensors_ = onnxruntime::make_unique<Provider_TensorProto_Impl>(v_.add_tensors());
-    return tensors_.get();
+struct Provider_Node__NodeIterator_Impl : Provider_Node__NodeIterator {
+  Provider_Node__NodeIterator_Impl(Node::NodeConstIterator&& v) : v_{std::move(v)} {}
+
+  bool operator!=(const Provider_Node__NodeIterator& p) const override { return v_ != static_cast<const Provider_Node__NodeIterator_Impl*>(&p)->v_; }
+
+  void operator++() override { v_.operator++(); }
+  const Provider_Node& operator*() override {
+    return *reinterpret_cast<const Provider_Node*>(&*v_);
   }
 
-  ONNX_NAMESPACE::AttributeProto v_;
-  std::unique_ptr<Provider_TensorProto_Impl> tensors_;
+  Node::NodeConstIterator v_;
 };
 
-struct Provider_KernelDef_Impl : Provider_KernelDef {
-  Provider_KernelDef_Impl(std::unique_ptr<KernelDef> p) : p_(std::move(p)) {}
-  std::unique_ptr<KernelDef> p_;
-};
+struct Provider_Node__EdgeIterator_Impl : Provider_Node__EdgeIterator {
```

Comment:
>_Node__EdgeIterator [](start = 50, length = 19)

We could simply nest them? The name is strange

Reply:
I wanted to nest them, but there's no way to nest prototypes. So in order to use the interior type before the ProviderHost interface, then have the definition later use both I had to do it this way.

```
struct Provider_Node;
struct Provider_Node::EdgeIterator; // I'd need something like this to do it.
```